### ISSUE TITLE
Entity-related events for the introspection module

### ISSUE THREAD

**Issue by Wendelstein7** [SquidDev-CC/ccjam-2016#177]
Today I came across an interesting idea which I'd like to see implemented into Plethora.

< tmi >
In a discussion regarding while loops I stand that a while loop is ultimately unneccecary in any case of use in ComputerCraft, as any changes in the environment (redstone, inventory, peripheral) fire events, so that any program can just have someone wait for events, rather than `while true checkLogic end`.

However, my conversation partner was talking about the interesting concept of measuring player damage, or more generalised, changes of entity-properties like hunger, health, potion effects, etc.
I did not think about this and quickly realised, this should have events!
< /tmi >

Therefore, I'd love to see events like the following to be introduced in Plethora, only available when the neural interface is equipped with the right modules:
- `health_changed` (on heal and damage)
- `hunger_changed`
- `saturation_changed`
- `potion_effect_induced`
- `potion_effect_ended`
- `xp_changed`
- `hand_use` (for when the player right clicks, like building blocks)
- `hand_destroy` (for when the player left clicks, like finishing destroying a block)
- `teleport` (for when the player used a portal and travveled dimensions)
- `jump` (for when the player jumps)
and more, but these were a few I could think of...

Be sure to talk about what you think about this!

With love
~HydroNitrogen

**Comment by kepler155c** [comment]
I would also like events for changes in the flags on getMetaOwner such as isSneaking and isElytraFlying.

**Comment by SquidDev** [comment]
Hydro and I chatted about this on Slack (well, I talked at them) when this was created:

The big issue here is that there aren't events for most of these - item usage, health and teleporting are pretty much the only things Forge provides hooks for. This means Plethora needs to check the entity every tick and track which fields have changed. There's a couple of reasons I'm not a fan of this:

 - **It's ugly:** I effectively need an "old entity state" class which holds a whole bunch of flags (previous hunger, previous position, etc...), then need to update that and monitor for changes.
 - **It's inefficient:** Even if it's opt-in/subscription based, you've now got to do a whole bunch of extra processing per tick.
 - **It's susceptible to scope creep:** As Kepler has demonstrated, there's always going to be another field someone wants to monitor. 

Honestly, it just seems easier (well, for me) and more flexible to poll the state every few ticks from Plethora. I realise this flies in the face of CC's event-driven design, but the alternative seems much worse.
